src/lib/profiles/data-management/Current/LoggingManagement.cpp

```diff
@@ -436,6 +437,210 @@ void LoggingManagement::DestroyLoggingManagement(void)
     Platform::CriticalSectionExit();
 }
 
+/**
+ * Persist the Weave events of all importance types.
+ *
+ * Serializes the events in WeaveCircularTLVBuffer and the associated states into the
+ * supplied buffer and set mShutdownInProgress flag to true such that no further
+ * events can be added to the buffer.
+ *
+ * This method is intended to be used by devices that do not retain RAM while sleeping,
+ * allowing them to persist events before going to sleep and thereby prevent lost of
+ * events
+ */
+WEAVE_ERROR LoggingManagement::PersistEvent(uint8_t * buf, uint32_t bufSize, uint32_t & serializedLen)
```

Comment:
This method should be named SerializeEvents() because its purpose is to serialize the currently stored events.  Whether the output will be persisted is immaterial.

Also, the method should take a reference to a TLVWriter.  This way the caller gets to decide where the data is encoded to.

Note that it is not very obvious how large the output should be.  Thus the caller will probably want to supply a TVLWriter that dynamically grows the underlying buffer.

Reply:
This is probably another dumb question - I guess when you init the TVLWriter with a fixed sized buffer, it's not going to grow? Could you provide me some key words for "TVLWriter that dynamically grows the underlying buffer", so I can search in codebase or google it? Thanks.

Reply:
Correct.  A TLVWriter initialized with a fixed buffer will fail if you try to write more than will fit.

I'm not sure where there is a good example of using a TLVWriter to write into a dynamic buffer.  We don't generally do this in openweave-core.    I can write you one if you want.

Reply:
Oh I see, thanks, I can write it myself~ 
Sorry I just thought there's some trick to use TLVWriter with fixed buffer to make it grow. I forgot there's dynamic buffer... really dumb...

Reply:
You might find it a little un-obvious how to do this, as it involves setting a custom GetNewBuffer callback on the TLVWriter object, and there aren't a lot of examples of that. 
 
There is documentation on the GetNewBuffer callback here: https://github.com/openweave/openweave-core/blob/efbed8f77e8c6a371508892abdfe3822f7aee14e/src/lib/core/WeaveTLVWriter.cpp#L69-L71

Let me know if you get stuck.

Reply:
Thank you very much.